lib/Controller/RegisterController.php

```diff
@@ -33,131 +42,218 @@ class RegisterController extends Controller {
+				[
+					'secret' => $secret,
+					'token' => $token,
+				]
+			)
+		);
+	}
 
-			return new TemplateResponse('registration', 'form', ['email' => $registration->getEmail(), 'token' => $registration->getToken()], 'guest');
-		} catch (RegistrationException $exception) {
-			return $this->renderError($exception->getMessage(), $exception->getHint());
+	/**
+	 * @NoCSRFRequired
+	 * @PublicPage
+	 *
+	 * @param string $secret
+	 * @param string $token
+	 * @return TemplateResponse
+	 */
+	public function showUserForm(string $secret, string $token, string $username = '', string $message = ''): TemplateResponse {
+		try {
+			$registration = $this->validateSecretAndToken($secret, $token);
+		} catch (RegistrationException $e) {
+			return $this->validateSecretAndTokenErrorPage();
 		}
+
+		return new TemplateResponse('registration', 'form/user', [
+			'email' => $registration->getEmail(),
+			'username' => $username,
+			'message' => $message,
+		], 'guest');
 	}
 
 	/**
 	 * @PublicPage
 	 * @UseSession
 	 *
-	 * @param $token
+	 * @param string $secret
+	 * @param string $token
+	 * @param string $username
+	 * @param string $password
 	 * @return RedirectResponse|TemplateResponse
+	 * @throws DoesNotExistException
 	 */
-	public function createAccount($token) {
-		$username = $this->request->getParam('username');
-		$password = $this->request->getParam('password');
-		$registration = $this->registrationService->getRegistrationForToken($token);
+	public function submitUserForm(string $secret, string $token, string $username, string $password) {
+		try {
+			$registration = $this->validateSecretAndToken($secret, $token);
+		} catch (RegistrationException $e) {
+			return $this->validateSecretAndTokenErrorPage();
+		}
 
 		try {
 			$user = $this->registrationService->createAccount($registration, $username, $password);
 		} catch (\Exception $exception) {
-			// Render form with previously sent values
-			return new TemplateResponse('registration', 'form',
-				[
-					'email' => $registration->getEmail(),
-					'entered_data' => ['user' => $username],
-					'errormsgs' => [$exception->getMessage()],
-					'token' => $token
-				], 'guest');
+			return $this->showUserForm($secret, $token, $username, $exception->getMessage());
 		}
 
 		if ($user->isEnabled()) {
+//			$appToken = $this->random->generate(72, ISecureRandom::CHAR_UPPER . ISecureRandom::CHAR_LOWER . ISecureRandom::CHAR_DIGITS);
+//
+//			$this->tokenProvider->generateToken(
+//				$appToken,
+//				$user->getUID(),
+//				$user->getUID(),
+//				$password,
+//				$this->getClientName(),
+//				IToken::PERMANENT_TOKEN
+//			);
+
+			\OC::$server->getUserSession()->login($user->getUID(), $password);
+			\OC::$server->getUserSession()->createSessionToken($this->request, $user->getUID(), $user->getUID(), $password);
+
+			/** @var LoginFlowV2Service $flowService */
+			$flowService = \OC::$server->get(LoginFlowV2Service::class);
+			$flowService->flowDone(
```

Comment:
Not really possible via public api, but i zuess this is okay, as it should really be an exception, right? @rullzer